bin/Macronify/Main.cpp

```diff
@@ -0,0 +1,734 @@
+
+#include <pasta/AST/AST.h>
+#include <pasta/AST/Decl.h>
+#include <pasta/Compile/Command.h>
+#include <pasta/Compile/Compiler.h>
+#include <pasta/Compile/Job.h>
+#include <pasta/Util/ArgumentVector.h>
+#include <pasta/Util/FileSystem.h>
+#include <pasta/Util/Init.h>
+
+#include <vast/Util/Common.hpp>
+#include <vast/Translation/CodeGenContext.hpp>
+#include <vast/Translation/CodeGenVisitor.hpp>
+#include <vast/Dialect/Dialects.hpp>
+#include <vast/Conversion/Passes.hpp>
+#include <vast/Translation/CodeGen.hpp>
+#include <vast/Translation/Register.hpp>
+#include <vast/Translation/CodeGenDriver.hpp>
+#include <vast/Translation/CodeGenTypeDriver.hpp>
+
+#include <macroni/Dialect/Macroni/MacroniDialect.hpp>
+#include <macroni/Dialect/Macroni/MacroniOps.hpp>
+#include <macroni/Dialect/Kernel/KernelDialect.hpp>
+#include <macroni/Dialect/Kernel/KernelOps.hpp>
+
+#include <cstdlib>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <optional>
+#include <algorithm>
+#include <set>
+#include <map>
+
+using Op = mlir::Operation;
+
+static const constexpr char get_user[] = "get_user";
+static const constexpr char offsetof[] = "offsetof";
+static const constexpr char container_of[] = "container_of";
+static const constexpr char rcu_read_lock[] = "rcu_read_lock";
+static const constexpr char rcu_read_unlock[] = "rcu_read_unlock";
+static const constexpr char rcu_dereference[] = "rcu_dereference";
+static const constexpr char smp_mb[] = "smp_mb";
+
+// TODO(bpp): Instead of using a global variable for the PASTA AST and MLIR
+// context, find out how to pass these to a CodeGen object.
+std::optional<pasta::AST> ast = std::nullopt;
+std::optional<mlir::MLIRContext> mctx = std::nullopt;
+std::optional<mlir::Builder> builder = std::nullopt;
+
+// TODO(bpp): Find a way to pass information along from the dynamic legality
+// callbacks to the conversion methods without using global variables
+std::map<Op *, Op *> get_user_to_x;
+std::map<Op *, Op *> get_user_to_ptr;
+
+std::map<Op *, mlir::TypeAttr> offsetof_to_type;
+std::map<Op *, mlir::StringAttr> offsetof_to_member;
+
+std::map<Op *, Op *> container_of_to_ptr;
+std::map<Op *, mlir::TypeAttr> container_of_to_type;
+std::map<Op *, mlir::StringAttr> container_of_to_member;
+
+std::map<Op *, Op *> list_for_each_to_pos;
+std::map<Op *, Op *> list_for_each_to_head;
+
+std::map<Op *, Op *> unlock_to_lock;
+
+std::map<Op *, Op *> rcu_dereference_to_p;
+
+static inline llvm::APInt get_lock_level(Op *op) {
+    return op->getAttrOfType<mlir::IntegerAttr>("lock_level").getValue();
+}
+
+namespace macroni {
+
+    // NOTE(bpp): It may be a good idea to also transform substitutions of
+    // get_user()'s parameters into special operations as well, to get even
+    // more information about the macro. This would let us match against all the
+    // various definitions of get_user, and all its substitutions of all its
+    // parameters.
+
+    struct macro_expansion_to_get_user
```

Comment:
I understand. I ask about functions because I saw that the `mlir::RewritePatternSet::add()` method is overloaded to also accept a function pointer, and was wondering if it would better to pass a function than a struct.